### ISSUE TITLE
Bug: element.current.scrollIntoView() not working as expected in chrome.

### ISSUE THREAD

**Issue by karthik2265** [facebook/react#23396]
Hey 🙋‍♂️

I have built a simple chatbot using react, and when the user sends a message i want to show the last message in the chatWindow component, so I used the following code:
```js
useEffect(
    function () {
      if (lastmessageRef.current !== null) {
        lastmessageRef.current.scrollIntoView()
      }
    },
    [lastmessageRef]
  )
```
It works as expected in edge and Firefox, but on chrome it is behaving weird
llink-to-the-chatbot: https://karthik2265.github.io/worlds-best-chatbot/
github-repo-link: https://github.com/karthik2265/worlds-best-chatbot

Thank you

**Comment by eps1lon** [comment]
Note that `lastmessageRef` as a dependency won't do anything and can be removed. The effect won't fire if `lastmessageRef.current` changes.

Could you include a video of the behavior in Firefox and Chrome and explain with words what's "weird" about this behavior?
Also please include the exact version of those browsers and the operating system you're using.

Bonus: Could you convert the repro into a codesandbox.io? Makes it easier to debug the issue.

**Comment by noviceGuru** [comment]
I'm having the same problem.
scrollIntoView api doesn't work inside useEffect, while it fires by onClick. Here is the code:

```
import React, { useEffect, useRef, useState } from 'react'
import FirstPage from './components/FirstPage'
import SecondPage from './components/SecondPage'
import ThirdPage from './components/ThirdPage'

import './App.css'

const App = () => {
  const [atPage, setAtPage] = useState<number>(0)
  const refs = useRef<(null | HTMLDivElement)[]>([])
  const pages = [<FirstPage />, <SecondPage />, <ThirdPage />]

  const handleWheel = (event: any) => {
    console.log(event.deltaY > 0 ? 'down' : 'up')
    setAtPage(atPage => {
      let atPageCopied = JSON.parse(JSON.stringify(atPage))
      return event.deltaY < 0 ? (atPage > 0 ? atPageCopied - 1 : atPage) : (atPage < pages.length - 1 ? atPageCopied + 1 : atPage)
    })
  }
  const scrollItThere = (atPage: number) => refs.current[atPage]?.scrollIntoView({ block: 'start', behavior: 'smooth' })

  useEffect(() => {
    console.log(atPage)
    scrollItThere(atPage)
  }, [atPage])


  return (
    <div onWheel={handleWheel} >
      {pages.map((el, index) => <div
        key={index}
        ref={i => refs.current[index] = i}
        onClick={() => refs.current[0]?.scrollIntoView({ block: 'start', behavior: 'smooth' })}
      >
        {el}
      </div>)}
    </div>
  )
}

export default App
```

**Comment by vigneshvjay** [comment]
> I'm having the same problem. scrollIntoView api doesn't work inside useEffect, while it fires by onClick. Worth mentioning that it works in Firefox. Here is the code:
> 
> ```
> import React, { useEffect, useRef, useState } from 'react'
> import FirstPage from './components/FirstPage'
> import SecondPage from './components/SecondPage'
> import ThirdPage from './components/ThirdPage'
> 
> import './App.css'
> 
> const App = () => {
>   const [atPage, setAtPage] = useState<number>(0)
>   const refs = useRef<(null | HTMLDivElement)[]>([])
>   const pages = [<FirstPage />, <SecondPage />, <ThirdPage />]
> 
>   const handleWheel = (event: any) => {
>     console.log(event.deltaY > 0 ? 'down' : 'up')
>     setAtPage(atPage => {
>       let atPageCopied = JSON.parse(JSON.stringify(atPage))
>       return event.deltaY < 0 ? (atPage > 0 ? atPageCopied - 1 : atPage) : (atPage < pages.length - 1 ? atPageCopied + 1 : atPage)
>     })
>   }
>   const scrollItThere = (atPage: number) => refs.current[atPage]?.scrollIntoView({ block: 'start', behavior: 'smooth' })
> 
>   useEffect(() => {
>     console.log(atPage)
>     scrollItThere(atPage)
>   }, [atPage])
> 
> 
>   return (
>     <div onWheel={handleWheel} >
>       {pages.map((el, index) => <div
>         key={index}
>         ref={i => refs.current[index] = i}
>         onClick={() => refs.current[0]?.scrollIntoView({ block: 'start', behavior: 'smooth' })}
>       >
>         {el}
>       </div>)}
>     </div>
>   )
> }
> 
> export default App
> ```

This works like a charm. Thanks buddy

**Comment by ValenW** [comment]
Come accross this issue, is there any progress?

**Comment by hwangyena** [comment]
I had the same problem
So I solved this by using setTimeout. (The flicker looks weird, but it works.)

```
// move to todo post
  useEffect(() => {
      ...
      const timer = setTimeout(() => {
      const todoPost = document.querySelector(`[data-post="${selectedTodoPost}"]`); 

      if (!todoPost) {
        return;
      }

      todoPost.scrollIntoView({ block: 'center' });
    }, 500);

    return () => clearTimeout(timer);
  }, [...]);
```

**Comment by jalexy12** [comment]
@hwangyena cool little thing I found out -- you can do what you did `setTimeout(..., 0)` and it works as well without delay, at least in my use case.

It's basically just a trick to defer until next tick in JS.

**Comment by eugenegodun** [comment]
I had the same problem and managed to fix it by using the `scroll` [method](https://developer.mozilla.org/en-US/docs/Web/API/Element/scroll) for the parent node instead of `scrollIntoView` for the children node:

`
const ref = useRef<HTMLDivElement>(null);

useEffect(() => {
        const scrollToTheBottom = () => {
            const scrollEl = ref.current;
            scrollEl?.scroll({
                top: scrollEl?.scrollHeight,
                behavior: 'smooth',
            });
        };
        scrollToTheBottom();
    }, []);

return (
        <Wrap>
            <Scroll ref={ref}>
                <MessageGroup messages={messagesIncome} />
                <MessageGroup messages={messagesIncome} />
                <MessageGroup messages={messagesIncome} />
                <MessageGroup messages={messagesIncome} />
                <MessageGroup messages={messagesIncome} />
                <MessageGroup messages={messagesIncome} />
                <MessageGroup messages={messagesIncome} />
                <MessageGroup messages={messagesIncome} />
                <MessageGroup messages={messagesIncome} />
                <MessageGroup messages={messagesIncome} />
                <MessageGroup messages={messagesIncome} />
                <MessageGroup messages={messagesIncome} />
                <MessageGroup messages={messagesIncome} />
                <MessageGroup messages={messagesIncome} />
                <MessageGroup messages={messagesIncome} />
                <MessageGroup messages={messagesIncome} />
                <MessageGroup messages={messagesIncome} />
            </Scroll>
        </Wrap>
    );
`

Works fine in Firefox, Chrome, Safari, Edge

https://user-images.githubusercontent.com/26645888/217844587-0b5d64d7-f34d-4128-8e43-9b7cefdc9403.mp4

**Comment by danielo515** [comment]
The problem doesn't seem to be limited to `lastmessageRef.current` if you fetch the component by ID, it also does not work on chrome, but works fine in firefox.

**Comment by ajoslin103** [comment]
Not working in Electron 19 (Chromium) 

This was working in React 16, I've since updated the project to React 18

**Comment by subhamayd2** [comment]
@pjm4 Yeah, that was the case for me as well.

This is what I did as a workaround:
```js
const [shouldScrollToRow, setShouldScrollToRow] = useState(false);

    useEffect(() => {
        if (shouldScrollToRow) {
            setTimeout(() => {
                elemRef.current?.scrollIntoView({ behavior: 'smooth', block: 'end' });
            }, 100);
            setShouldScrollToRow(false);
        }
    }, [shouldScrollToRow]);

    useEffect(() => {
        // Listening for items getting added
        if (itemsUpdated) {
            setShouldScrollToRow(true);
            setItemsUpdated(false);
        }
    // eslint-disable-next-line react-hooks/exhaustive-deps
    }, [itemsUpdated]);
```

**Comment by Jerry-qiqi** [comment]
I want to know why this is happening

**Comment by jordanl17** [comment]
A workaround I've found it to use `window.requestAnimationFrame`:

```
useEffect(() => {
  if (itemsUpdated) {
    window.requestAnimationFrame(() => 
      elRef.current?.scrollIntoView()
    )
  }
}, [])
```

**Comment by mateussilva92** [comment]
Just wanted to shed some light  into this issue and why it can probably be closed.

This has nothing to do with react. This is a problem that is related to the chromium implementation of the `scrollIntoView` method.
There's been some open issues for a long time in the chromium project regarding this issue.

https://bugs.chromium.org/p/chromium/issues/detail?id=1121151
https://bugs.chromium.org/p/chromium/issues/detail?id=1043933
https://bugs.chromium.org/p/chromium/issues/detail?id=833617

So from what I've got from all the testing and searching the last few days is that chromium based browsers interrupt the `scrollIntoView` when it any other scroll event is triggered.

This only happens with the `scrollIntoView` method, the `scrollTo` and `scroll` methods don't suffer from the this issue.
You can see and test this by yourself on the following fiddle.

https://jsfiddle.net/2bnspw8e/8/

So then what's a solution if you want to keep the simplicity of `scrollIntoView` and have other scroll happen at the same time?

There is a [**ponyfill**](https://github.com/sindresorhus/ponyfill) version of this [method](https://github.com/scroll-into-view/scroll-into-view-if-needed) which reimplements the method using the non broken `scroll` method.

**Comment by ajoslin103** [comment]
Thanks for the digging and, I always wanted a one-trick pony! 

This shows up at just the right time!

Cheers!

**Comment by bn1ck** [comment]
I ended up creating my own helper function.

```
export const smoothScroll = (targetY: number, duration: number) => {
  const startingY = window.pageYOffset;
  const diff = targetY - startingY;
  let start: number;

  window.requestAnimationFrame(function step(timestamp) {
    if (!start) start = timestamp;
    const time = timestamp - start;
    const percent = Math.min(time / duration, 1);

    window.scrollTo(0, startingY + diff * percent);

    if (time < duration) {
      window.requestAnimationFrame(step);
    }
  });
}
```
To call the function i use the following 


```
if (viewerRef.current) {
      const top = viewerRef.current.offsetTop;
      smoothScroll(top, 500);
}
```

**Comment by iuriikomarov** [comment]
IMO almost all the answers above are not correct.

You are trying to invoke scrolling in `Effect` but it should be done in `LayoutEffect`.

Function inside effect executes in the point in parallel to DOM update commits and in 99% it will be run faster. So normally real DOM (according to the vDOM built up with your render function) is not ready yet at that point of time. This is why "workarounds" using `setTimeout` work somehow. But [using `LayoutEffect` is more accurate](https://react.dev/reference/react/useLayoutEffect).

**Comment by danielo515** [comment]
this is not stale, because it is not fixed

**Comment by hypo-thesis** [comment]
This is not fixed for sure. `behavior: 'smooth'` does work on FF but not on Chrome. I have codes to reproduce if you wish.

**Comment by missalyss** [comment]
I want to add that the `useRef`, `useEffect` combo didn't work for me (no movement at all, because `elementRef.current` wasn't defined until after the effect had run). Instead, I updated to `useCallback`.

```
  const elementRef = useCallback((domEl) => {
    if (domEl !== null ) {
      domEl.scrollIntoView({
        behavior: 'smooth',
      });
    }
  }, []);

...

<div ref={elementRef} >...</div>
```

**Comment by alexpeev9** [comment]
Hello. 

After checking different scenarios, I also concluded that there is an issue with the way Chrome behaves with `scrollIntoView()` when you set the behavior to `smooth`. 

In my case, the scroll randomly stops at the start or middle of the scroll. 
A **temporary** solution for me would be to disable the smooth scroll for Chrome browsers.

```
const behavior = navigator.userAgent.includes('Chrome') ? 'auto' : 'smooth'
ref.current?.scrollIntoView({ behavior: behavior, block: 'start' })
```

**Comment by Qnemes** [comment]
In my case I`ve encountered weird behavior with simple onPress event with window.scrollTo({ top: scrollRef.current.offsetTop, behavior: 'smooth' }); So I've wrapped this call with setTimeout 0ms and it fixed unexpected behavior (stops mid scroll) completely.

**Comment by subzeta** [comment]
> In my case I`ve encountered weird behavior with simple onPress event with window.scrollTo({ top: scrollRef.current.offsetTop, behavior: 'smooth' }); So I've wrapped this call with setTimeout 0ms and it fixed unexpected behavior (stops mid scroll) completely.

This was the issue in my case. Thanks mate!

**Comment by satejbidvai** [comment]
Any updates on this?

I still seem to be getting this exact issue on Chrome. Removing `behaviour: smooth` works.

**Comment by mmis1000** [comment]
> Any updates on this?
> 
> I still seem to be getting this exact issue on Chrome. Removing `behaviour: smooth` works.

You should probably look into related chrome issue instead. This is a chrome bug that has nothing to do with react. 

https://bugs.chromium.org/p/chromium/issues/detail?id=833617

The recent comment says thet are changing how smooth in to view smooth work in chrome, which may fix this issue.

**Comment by scath999** [comment]
Still not working for me in either Chrome or Firefox on desktop.

I enclose an excerpt from some messaging functionality - users can open and close a chat window from a list of available conversations.

The first time I open the window, the content appears but the window doesn't scroll. If I then close and reopen the window it works as intended, as does posting new text to the window.

Note: The placement of scrollIntoView() is in two different places within these methods - this is intentionally done to test the behaviour; afaik both should work, neither does.

```
     const scroll_position = useRef()
 
     (...) 
     
     function GetMessageThread(thread_id){
  
        axios.post("/api/get_message_thread", {
            headers: {
                Accept: 'application/json',
                'Content-Type': 'application/json',
                withCredentials: true
            },
            body: JSON.stringify({"thread_id" : thread_id})
        })
        .then((response) => {
            if (response.statusText == "OK"){
                setMessageList(response.data['message_list'])
                console.log(":: GetMessageThread :: Scrolling to bottom of window ")
                console.log("-- scroll_position.current =", scroll_position.current)
                scroll_position.current.scrollIntoView()
              
            }
        })
    }
    
   

    function NewMessage(e){
        e.preventDefault();

        var payload = {"message" : newMessage, "thread_id" : thread_id}
        axios.post('/api/new_message', {
            headers: {
                Accept: 'application/json',
                'Content-Type': 'application/json',
                withCredentials: true
            },
            body: JSON.stringify(payload)
        }).then((response) =>{
            if (response.statusText == "OK"){
               console.log("response from new_message = OK")
              
            }
        })

        setNewMessage('')
        console.log(":: NewMessage :: Scrolling to bottom of window")
        scroll_position.current.scrollIntoView({ behavior : 'smooth'})

    }
    
    return(
       <>
        (...)
         <DialogContent>            
                {messageList.map((msg, i) => 
                    {
                        return (<ChatMessage key={i} messageObj={msg}  />)
                    }
                )}
                <span id="scroll_pos" name="scroll_pos" ref={scroll_position}></span>
            </DialogContent>
        (...)
        </>
    )
```

Output in console, whether on first or subsequent opens, is as follows :

```
:: GetMessageThread :: Scrolling to bottom of window 
-- scroll_position.current = <span id=​"scroll_pos" name=​"scroll_pos">​</span>​
```

**Comment by himself65** [comment]
I'm having the similar issue, but even serious bug on next.js 14.2.4 where document.body will have a werid ";" content

<img width="677" alt="image" src="https://github.com/user-attachments/assets/2d8c6ac7-23fa-4b39-9cec-655cbf61b170">

And the code was

```ts
  useEffect(() => {
    if (trackVisibility && !inView) {
      entry?.target.scrollIntoView({
        block: "start",
      });
    }
  }, [entry?.target, inView, trackVisibility]);
```

and I change to no use effect and it will fix the behavior

```ts
  const onceRef = useRef(false);
  if (!onceRef.current) {
    if (trackVisibility && !inView) {
      entry?.target.scrollIntoView({
        block: "start",
      });
    }
    onceRef.current = true;
  }
```

I want to make a mininum reproduce code but unfortually I didn't make it.

Nore sure if @eps1lon have idea of this bug

**Comment by himself65** [comment]
<img width="830" alt="image" src="https://github.com/user-attachments/assets/40416bfa-70c8-4bd9-955f-91f3c80a13d9">

set debugger before `entry?.target.scrollIntoView` and I found that filber id is same one

<img width="820" alt="image" src="https://github.com/user-attachments/assets/2f280bab-ecbd-4ca0-9f21-664dc0301949">
<img width="577" alt="image" src="https://github.com/user-attachments/assets/64150bb8-4184-4f1c-a515-f40f43cdfbef">

**Comment by Kodiererin** [comment]
> Still not working for me in either Chrome or Firefox on desktop.
> 
> I enclose an excerpt from some messaging functionality - users can open and close a chat window from a list of available conversations.
> 
> The first time I open the window, the content appears but the window doesn't scroll. If I then close and reopen the window it works as intended, as does posting new text to the window.
> 
> Note: The placement of scrollIntoView() is in two different places within these methods - this is intentionally done to test the behaviour; afaik both should work, neither does.
> 
> ```
>      const scroll_position = useRef()
>  
>      (...) 
>      
>      function GetMessageThread(thread_id){
>   
>         axios.post("/api/get_message_thread", {
>             headers: {
>                 Accept: 'application/json',
>                 'Content-Type': 'application/json',
>                 withCredentials: true
>             },
>             body: JSON.stringify({"thread_id" : thread_id})
>         })
>         .then((response) => {
>             if (response.statusText == "OK"){
>                 setMessageList(response.data['message_list'])
>                 console.log(":: GetMessageThread :: Scrolling to bottom of window ")
>                 console.log("-- scroll_position.current =", scroll_position.current)
>                 scroll_position.current.scrollIntoView()
>               
>             }
>         })
>     }
>     
>    
> 
>     function NewMessage(e){
>         e.preventDefault();
> 
>         var payload = {"message" : newMessage, "thread_id" : thread_id}
>         axios.post('/api/new_message', {
>             headers: {
>                 Accept: 'application/json',
>                 'Content-Type': 'application/json',
>                 withCredentials: true
>             },
>             body: JSON.stringify(payload)
>         }).then((response) =>{
>             if (response.statusText == "OK"){
>                console.log("response from new_message = OK")
>               
>             }
>         })
> 
>         setNewMessage('')
>         console.log(":: NewMessage :: Scrolling to bottom of window")
>         scroll_position.current.scrollIntoView({ behavior : 'smooth'})
> 
>     }
>     
>     return(
>        <>
>         (...)
>          <DialogContent>            
>                 {messageList.map((msg, i) => 
>                     {
>                         return (<ChatMessage key={i} messageObj={msg}  />)
>                     }
>                 )}
>                 <span id="scroll_pos" name="scroll_pos" ref={scroll_position}></span>
>             </DialogContent>
>         (...)
>         </>
>     )
> ```
> 
> Output in console, whether on first or subsequent opens, is as follows :
> 
> ```
> :: GetMessageThread :: Scrolling to bottom of window 
> -- scroll_position.current = <span id=​"scroll_pos" name=​"scroll_pos">​</span>​
> ```

I also faced this same issue, had to create my own custom function,

**Comment by JulioMacedo0** [comment]
In my use case, the following code snippet worked very well:

```javascript
divRef.current?.scroll({
  top: divRef.current?.scrollHeight,
  behavior: 'smooth'
});

**Comment by say8425** [comment]
Yes, I know it's a `setTimeout`, but works more accurately.

```ts
useEffect(() => {
  if (itemsUpdated) {
    window.requestAnimationFrame(() => {
      setTimeout(() => {
        ref.current?.scrollIntoView();
      }, 100);
    });
  }
}, []);
```

**Comment by lulucassiu2** [comment]
lucassiu2@outlook.my

**Comment by HureraNadeem** [comment]
```
 useEffect(() => {
    if (element.current) {
      const scrollElement = element.current.contentWrapperEl;
      scrollElement.scrollTop = scrollElement.scrollHeight;
    }
}, []);

<div ref={chatWindow} />
```

This worked for me, could find no way to access `scrollIntoView()` or `scroll()` from `element.current`

**Comment by ludwig-gramberg** [comment]
I ran into the same issue but in another context with Alpinejs.

Observation 1:

You would think, based on your code/framework, that the DOM has finished rendering and it is safe to scroll to an element, because it is visible. But sometimes the browser does not do it.

One example was a `<ul>` being made visible only once it had elements in it. The framework indicated that it had finished rendering but the browser wasn't quite there yet, perhaps a repaint did not actually finish yet? So I think this is some sort of -race-condition within the browser-engines where sometimes it is ready and sometimes it is not. In my example if the `<ul>` was always visible the scroll always worked.
This is why a setTimeout helped. The problem is we do not know what timeout is required on which device/engine.

Observation 2: 

Using the MutationObserver the browsers gave me reliable callbacks when stuff was truly rendered/finished. Scrolls made here always worked. The only problem was that it is hard to generalize this approach.

**Comment by aasitvora99** [comment]
Just chiming into the conversation. I used webdriver's scrollIntoView and that doesn't work well with react-native and ionic app frameworks.

**Comment by tomo-kn** [comment]
We're experiencing a similar issue in iframe/widget environments. 
When using scrollIntoView({ block: 'start' }), our iframe widget gets visually shifted upward.
This doesn't happen with block: 'center', block: 'end', or block: 'nearest'.

Switching to block: 'nearest' resolves the positioning issue.
Seems like iframe coordinate calculations make block: 'start' more sensitive to timing issues between React updates and browser scroll positioning.

**Comment by typescript-class850** [comment]
The solution’s clear https://kingston-layover.github.io/restlet/
